File: native_crypto_ios/ios/Classes/Cipher.swift
/**
 * Author: Hugo Pointcheval
 * Email: git@pcl.ovh
 * -----
 * File: Cipher.swift
 * Created Date: 25/12/2021 18:31:28
 * Last Modified: 25/12/2021 18:38:53
 * -----
 * Copyright (c) 2021
 */

import Foundation
import CryptoKit

class AESCipher {
    /// Encrypts plaintext with key using AES GCM
    @available(iOS 13.0, *)
    static func encrypt(plaintext: Data, key: Data) -> Data? {
        let symmetricKey = SymmetricKey.init(data: key)
        let encrypted = try? AES.GCM.seal(plaintext, using: symmetricKey)
        return encrypted?.combined
    }
    
    /// Decrypts ciphertext with key using AES GCM
    @available(iOS 13.0, *)
    static func decrypt(ciphertext: Data, key: Data) -> Data? {
        let symmetricKey = SymmetricKey.init(data: key)
        let sealedBox = try? AES.GCM.SealedBox(combined: ciphertext)
        if (sealedBox == nil) { return nil }
        let decryptedData = try? AES.GCM.open(sealedBox!, using: symmetricKey)
        return decryptedData
    }
}

class CHACHACipher {
    /// Encrypts plaintext with key using CHACHAPOLY
    @available(iOS 13.0, *)
    static func encrypt(plaintext: Data, key: Data) -> Data? {
        let symmetricKey = SymmetricKey.init(data: key)
        let encrypted = try? ChaChaPoly.seal(plaintext, using: symmetricKey)
        return encrypted?.combined
    }
    
    /// Decrypts ciphertext with key using CHACHAPOLY
    @available(iOS 13.0, *)
    static func decrypt(ciphertext: Data, key: Data) -> Data? {
        let symmetricKey = SymmetricKey.init(data: key)
        let sealedBox = try? ChaChaPoly.SealedBox(combined: ciphertext)
        if (sealedBox == nil) { return nil }
        let decryptedData = try? ChaChaPoly.open(sealedBox!, using: symmetricKey)
        return decryptedData
    }
}

File: native_crypto/lib/src/cipher.dart
// Author: Hugo Pointcheval
// Email: git@pcl.ovh
// -----
// File: cipher.dart
// Created Date: 16/12/2021 16:28:00
// Last Modified: 28/12/2021 12:25:38
// -----
// Copyright (c) 2021

import 'dart:typed_data';

import 'cipher_text.dart';

/// Represents different cipher algorithms
enum CipherAlgorithm { aes, rsa }

/// Represents a cipher.
///
/// In cryptography, a cipher is an algorithm for performing encryption
/// or decryption - a series of well-defined steps that can
/// be followed as a procedure.
abstract class Cipher {
  /// Returns the standard algorithm name for this cipher
  CipherAlgorithm get algorithm;

  /// Encrypts data.
  ///
  /// Takes [Uint8List] data as parameter.
  /// Returns a [CipherText].
  Future<CipherText> encrypt(Uint8List data);

  /// Decrypts cipher text.
  ///
  /// Takes [CipherText] as parameter.
  /// And returns plain text data as [Uint8List].
  Future<Uint8List> decrypt(CipherText cipherText);
}